### PR TITLE
POWER back-end: wrong compilation of FP32 stores

### PR DESCRIPTION
**Pull request by xavierleroy**
Amusingly, I merged two PRs today, and the second PR (#10775) contains a new test that reveals a bug in the first PR (#12276)...

The bug, which has been present since OCaml 1.00, is that wrong POWER code is emitted for stores of 32-bit FP numbers.

In the POWER architecture, FP32 numbers are stored in FP registers in FP64 format, so that FP32 -> FP64 conversions are no-ops.  However, the `stfs` instruction for FP32 stores assumes its argument to be a FP32 number in FP64 format, not any FP64 number.  So, an explicit FP64 -> FP32 rounding was missing before `stfs`.

This PR adds the missing rounding (`frsp`).  Since a temporary FP register is needed, the PR also clarifies the  status of FP register 0, making it accessible as `phys_reg 100` and, as a consequence, available for register allocation.

### CODE REVIEW

File: asmcomp/power/proc.ml [hunk 1]
```diff
@@ -79,7 +79,7 @@ let hard_int_reg =
   for i = 0 to 22 do v.(i) <- Reg.at_location Int (Reg i) done; v
 
 let hard_float_reg =
-  let v = Array.make 31 Reg.dummy in
+  let v = Array.make 32 Reg.dummy in
   for i = 0 to 30 do v.(i) <- Reg.at_location Float (Reg(100 + i)) done; v
```
Comment (avsm) [inline asmcomp/power/proc.ml]: Doesn't this also need to be bumped to `0 to 31` here?

Comment (xavierleroy) [reply asmcomp/power/proc.ml]: Definitely!  This is now fixed.

File: asmcomp/power/proc.ml [hunk 2]
```diff
@@ -264,7 +267,7 @@ let safe_register_pressure = function
 
 let max_register_pressure = function
     Iextcall _ -> [| 13; 18 |]
-  | _ -> [| 23; 30 |]
+  | _ -> [| 23; 32 |]
```
Comment (avsm) [inline asmcomp/power/proc.ml:271]: Was this `30` incorrect before, as there has only been one register added to the allocation pool in this PR? (Just asking for my own understanding)

**Review by avsm — APPROVED** [review]
Tested on FreeBSD/powerpc64le and Linux/ppc64le;  code looks right to me.

File: asmcomp/power/proc.ml (hunk 2 above)
Comment (xavierleroy) [reply asmcomp/power/proc.ml:271]: Right.  I thought for a minute that 30 was a leftover from a time when there was only 30 allocatable FP registers, but it was not.  30 (before) / 31 (now) is conservative but correct, as it accounts for the temporary register used for intoffloat and for store float32.  32 (now) does not account for that, my mistake.  I just pushed a more explicit version of this function that distinguishes between the intoffloat/store-float32 cases and the other instructions.

File: asmcomp/power/emit.mlp [hunk 3]
```diff
@@ -678,14 +683,18 @@ let emit_instr env i =
...
+        emit_load_store "stfs" addr i.arg 1 tmp
     | Lop(Istore(chunk, addr, assignment)) ->
         let storeinstr =
           match chunk with
           | Byte_unsigned | Byte_signed -> "stb"
           | Sixteen_unsigned | Sixteen_signed -> "sth"
           | Thirtytwo_unsigned | Thirtytwo_signed -> "stw"
           | Word_int | Word_val -> "std"
-          | Single -> "stfs"
+          | Single -> assert false
```
Comment (tmcgilchrist) [inline asmcomp/power/emit.mlp:697]: Just to clarify from my reading this branch is unreachable because of the special case added above.

**Review by tmcgilchrist — COMMENTED** [review]
Tested on Linux/ppc64le (Power9), the code looks right to me as well.

From the Power ISA v3.0B manual
> 4. Store Floating-Point Single
This form of instruction converts a double-precision operand to single format and stores that oper and into storage. No floating-point exceptions are caused by these instructions. (The value being stored is effectively assumed to be the result of an instruction of one of the preceding three types.)

**Comment by xavierleroy** [comment]
Thanks all for  the reviews, and sorry for the back-and-forth on the "make FPR0 allocatable" part.

**Pull request merged**